From 3a1d6d3e2bd0e1506995e316b183cef8abc0034b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Christoph=20B=C3=B6hmwalder?=
 <christoph.boehmwalder@linbit.com>
Date: Fri, 6 Sep 2024 11:14:43 +0200
Subject: [PATCH 26/32] compat: test and patch for __counted_by

Unfortunately, the "attribute name" variable type only works with very
recent coccinelle versions (>=1.1.1). So add a hint about that in the
README.
---
 README.md                                            |  9 ++++-----
 .../cocci/counted_by__no_present.cocci               | 12 ++++++++++++
 drbd/drbd-kernel-compat/gen_patch_names.c            |  3 +++
 drbd/drbd-kernel-compat/tests/have_counted_by.c      |  8 ++++++++
 4 files changed, 27 insertions(+), 5 deletions(-)
 create mode 100644 drbd/drbd-kernel-compat/cocci/counted_by__no_present.cocci
 create mode 100644 drbd/drbd-kernel-compat/tests/have_counted_by.c

diff --git a/README.md b/README.md
index ec513d19438e..b63286db0113 100644
--- a/README.md
+++ b/README.md
@@ -38,7 +38,7 @@ Since version 9.0.20, DRBD has been using a kernel backwards compatibility syste
 based on [Coccinelle](https://github.com/coccinelle/coccinelle) semantic patches.
 While this has many advantages, it also makes it a little harder for "casual"
 developers to build DRBD from the git sources. The problem is that we require a
-very recent version of `spatch` (at least 1.0.8 at time of writing), and most
+very recent version of `spatch` (at least 1.1.1 at time of writing), and most
 distributions only have relatively old versions in their repositories.
 
 ## From git
@@ -48,10 +48,9 @@ For users wishing to build DRBD from its git sources, here are a few options:
    chances are you won't even have to use any of the compat features, which
    means you won't require compatibility patches and in turn don't need spatch
    installed.
-2. On Ubuntu 18.04 and newer, use a recent spatch version from the
-   [Coccinelle PPA](https://launchpad.net/~npalix/+archive/ubuntu/coccinelle).
-   This provides (at time of writing) version 1.0.8, which is recent enough to
-   build DRBD.
+2. Fedora >=39 packages Coccinelle 1.1.1, which is recent enough to build DRBD. **Warning**: while
+   Ubuntu also ships Coccinelle with supposedly the same version number (1.1.1), this version
+   is *not* able to build DRBD.
 3. Build and install spatch from source. This will also give you a version that
    is recent enough to build DRBD.
 
diff --git a/drbd/drbd-kernel-compat/cocci/counted_by__no_present.cocci b/drbd/drbd-kernel-compat/cocci/counted_by__no_present.cocci
new file mode 100644
index 000000000000..40db31bece0c
--- /dev/null
+++ b/drbd/drbd-kernel-compat/cocci/counted_by__no_present.cocci
@@ -0,0 +1,12 @@
+@@
+type T;
+identifier s, x;
+attribute name __counted_by;
+@@
+struct s {
+...
+	T x
+-	__counted_by(...)
+	;
+...
+};
diff --git a/drbd/drbd-kernel-compat/gen_patch_names.c b/drbd/drbd-kernel-compat/gen_patch_names.c
index cc3bad2d84e4..72c1e25512ea 100644
--- a/drbd/drbd-kernel-compat/gen_patch_names.c
+++ b/drbd/drbd-kernel-compat/gen_patch_names.c
@@ -605,6 +605,9 @@ int main(int argc, char **argv)
 	patch(1, "need_recalc_sigpending", false, true,
 		COMPAT_NEED_RECALC_SIGPENDING, "need");
 
+	patch(1, "counted_by", true, false,
+		COMPAT_HAVE_COUNTED_BY, "present");
+
 /* #define BLKDEV_ISSUE_ZEROOUT_EXPORTED */
 /* #define BLKDEV_ZERO_NOUNMAP */
 
diff --git a/drbd/drbd-kernel-compat/tests/have_counted_by.c b/drbd/drbd-kernel-compat/tests/have_counted_by.c
new file mode 100644
index 000000000000..07fac32aad16
--- /dev/null
+++ b/drbd/drbd-kernel-compat/tests/have_counted_by.c
@@ -0,0 +1,8 @@
+/* { "version": "v6.4", "commit": "dd06e72e68bcb4070ef211be100d2896e236c8fb", "comment": "add __counted_by macro", "author": "Kees Cook <keescook@chromium.org>", "date": "Wed May 17 12:08:44 2023 -0700" } */
+
+#include <linux/compiler_attributes.h>
+
+struct foo {
+	int a;
+	int b[] __counted_by(a);
+};
-- 
2.35.3

